### PR TITLE
Implementing Blur filter

### PR DESCRIPTION
**Pull request by Adenilson**
See discussion on #5190 and #5496.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/5546)

### CODE REVIEW

**Comment by Adenilson** [comment]
@SimonSapin and @pcwalton 

I have one question about paint_context + StackingContext. 

As was commented, it might be needed to adjust the overflow regions or (quote) “else display list optimization might clip out the edges of a blurred object”. 

But reading the code, I noticed that the filters are only draw (e.g. http://mxr.mozilla.org/servo/source/components/gfx/paint_context.rs#923) way past the display list optimizer has executed (i.e. http://mxr.mozilla.org/servo/source/components/gfx/display_list/mod.rs#383). 

If so, that would require changes on optimize_and_draw_into_context() i.ie to take into account the fact that there are filters *before* performing the display list optimization?

**Comment by Adenilson** [comment]
Testing with the following test case (used 512px as IIRC, that is the default tile size): https://gist.github.com/Adenilson/ae28080d9fc25bb64ff1

Got attached result. The clipping probably is responsible for the smaller size, but what about the white lines in the green rectangle?

ff_blur
servo_blur

**Comment by pcwalton** [comment]
@Adenilson That indicates that your intermediate draw target sizes are incorrect. The white lines are on tile boundaries (the sizes of which can be changed with the `-s` flag).

**Comment by pcwalton** [comment]
@Adenilson Looking at the `box-shadow` painting code may help.

**Comment by Adenilson** [comment]
WIP.
screen shot 2015-04-10 at 10 33 12 am

**Comment by maxhoffmann** [comment]
Some time ago I saw this video about color blurring being broken. Not sure if this is the right thread to mention it, but maybe Servo could be the first rendering engine to blur colors correctly? :)

https://www.youtube.com/watch?v=LKnqECcg6Gw

**Comment by SimonSapin** [comment]
@maxhoffmann We (Servo) will not unilaterally implement a behavior incompatible with other browsers, however more correct it might be. This is a proposal for W3C.

That said, is [the `color-interpolation-filters` property](http://dev.w3.org/fxtf/filters/#propdef-color-interpolation-filters) doing this already?

File: components/style/properties.mako.rs [hunk 1]
```diff
@@ -2798,6 +2777,44 @@ pub mod longhands {
...
+                    try!(filter.to_css(dest));
+                }
+                Ok(())
+            }
+        }
+
+        impl ToCss for SpecifiedFilter {
+            fn to_css<W>(&self, dest: &mut W) -> text_writer::Result where W: TextWriter {
+                match *self {
+                    SpecifiedFilter::Blur(value) => try!(write!(dest, "blur({:?})", value)),
```
Comment (SimonSapin) [inline components/style/properties.mako.rs]: `{:?}` is not necessarily doing the right formatting here. Use `value.to_css(dest)` instead, like `hue-rotate()`. I made a PR into your PR’s branch https://github.com/Adenilson/servo/pull/1. Merging it should automatically update this one.

**Comment by SimonSapin** [comment]
r=me on the `components/style` changes if https://github.com/Adenilson/servo/pull/1 is included.

**Comment by maxhoffmann** [comment]
@SimonSapin As far as I understand the video, the calculations of the blended colors are calculated wrongly independent of the color space, so the mentioned property wouldn’t fix that. Thanks for the quick answer.

**Comment by SimonSapin** [comment]
“Square root everything” sounds a lot like a color space, but I don’t know. Regardless, rather than Servo just deviating from the spec, the right way to make this happen is to first discuss at W3C to change the spec. See “Feedback” at the top of http://dev.w3.org/fxtf/filters/

**Comment by Adenilson** [comment]
Related #5668.

File: components/gfx/paint_context.rs [hunk 2]
```diff
@@ -909,12 +909,37 @@ impl<'a> PaintContext<'a> {
...
         // FIXME(pcwalton): This surface might be bigger than necessary and waste memory.
-        let size = self.draw_target.get_size();
-        let size = Size2D(size.width, size.height);
+        let size = self.draw_target.get_size(); //Az size.
+        let mut size = Size2D(size.width, size.height); //Geom::Size.
+
+        // Pre-calculate if there is a blur expansion need.
+        let accum_blur = filters::calculate_accumulated_blur(filters);
+        let mut matrix = Matrix2D::identity();
+        if accum_blur > Au::new(0) {
```
Comment (pcwalton) [inline components/gfx/paint_context.rs]: nit: I prefer just `Au(0)`

File: components/gfx/paint_context.rs [hunk 3]
```diff
@@ -909,12 +909,37 @@ impl<'a> PaintContext<'a> {
...
+            let temporary_draw_target_bounds = old_transform.transform_rect(&inflated_size);
+            matrix = Matrix2D::identity().translate(-temporary_draw_target_bounds.origin.x as AzFloat,
+                                                    -temporary_draw_target_bounds.origin.y as AzFloat).mul(&old_transform);
+        }
 
         let temporary_draw_target =
             self.draw_target.create_similar_draw_target(&size, self.draw_target.get_format());
-        temporary_draw_target.set_transform(&self.draw_target.get_transform());
+
+        if accum_blur > Au::new(0) {
```
Comment (pcwalton) [inline components/gfx/paint_context.rs]: same here

File: components/gfx/paint_context.rs [hunk 4]
```diff
@@ -909,12 +909,37 @@ impl<'a> PaintContext<'a> {
...
 
         // FIXME(pcwalton): This surface might be bigger than necessary and waste memory.
-        let size = self.draw_target.get_size();
-        let size = Size2D(size.width, size.height);
+        let size = self.draw_target.get_size(); //Az size.
+        let mut size = Size2D(size.width, size.height); //Geom::Size.
+
+        // Pre-calculate if there is a blur expansion need.
+        let accum_blur = filters::calculate_accumulated_blur(filters);
+        let mut matrix = Matrix2D::identity();
```
Comment (pcwalton) [inline components/gfx/paint_context.rs]: Why not just use `let mut matrix = self.draw_target.get_transform();`? Then you wouldn't need the second `if accum_blur > Au(0) { ...` below.

File: components/gfx/paint_context.rs [hunk 5]
```diff
@@ -932,18 +957,35 @@ impl<'a> PaintContext<'a> {
...
+        let rect = Rect(Point2D(0.0, 0.0), self.draw_target.get_size().to_azure_size());
+
+        let rect_temporary = Rect(Point2D(0.0, 0.0), temporary_draw_target.get_size().to_azure_size());
 
         // Create the Azure filter pipeline.
+        let mut accum_blur = Au::new(0);
         let (filter_node, opacity) = filters::create_filters(&self.draw_target,
                                                              temporary_draw_target,
-                                                             filters);
+                                                             filters, &mut accum_blur);
```
Comment (pcwalton) [inline components/gfx/paint_context.rs]: nit: line up this argument with the others

File: components/gfx/paint_context.rs [hunk 6]
```diff
@@ -932,18 +957,35 @@ impl<'a> PaintContext<'a> {
         // Set up transforms.
         let old_transform = self.draw_target.get_transform();
         self.draw_target.set_transform(&Matrix2D::identity());
-        temporary_draw_target.set_transform(&Matrix2D::identity());
+        let rect = Rect(Point2D(0.0, 0.0), self.draw_target.get_size().to_azure_size());
+
+        let rect_temporary = Rect(Point2D(0.0, 0.0), temporary_draw_target.get_size().to_azure_size());
 
         // Create the Azure filter pipeline.
+        let mut accum_blur = Au::new(0);
```
Comment (pcwalton) [inline components/gfx/paint_context.rs]: nit: `Au(0)`

File: components/gfx/paint_context.rs [hunk 7]
```diff
@@ -932,18 +957,35 @@ impl<'a> PaintContext<'a> {
...
+                                                             filters, &mut accum_blur);
 
         // Perform the blit operation.
-        let rect = Rect(Point2D(0.0, 0.0), self.draw_target.get_size().to_azure_size());
         let mut draw_options = DrawOptions::new(opacity, 0);
         draw_options.set_composition_op(blend_mode.to_azure_composition_op());
-        self.draw_target.draw_filter(&filter_node, &rect, &rect.origin, draw_options);
+
+       // If there is a blur expansion, shift the transform and update the size.
+        if accum_blur > Au::new(0) {
```
Comment (pcwalton) [inline components/gfx/paint_context.rs]: nit: `Au(0)`

File: components/gfx/paint_context.rs [hunk 8]
```diff
@@ -909,12 +909,37 @@ impl<'a> PaintContext<'a> {
...
+
+        // Pre-calculate if there is a blur expansion need.
+        let accum_blur = filters::calculate_accumulated_blur(filters);
+        let mut matrix = Matrix2D::identity();
+        if accum_blur > Au::new(0) {
+            // Set the correct size.
+            let rect = Rect(Point2D(0, 0), size);
+            let side_inflation = accum_blur * BLUR_INFLATION_FACTOR;
+            let inflated_size = rect.inflate(side_inflation.to_nearest_px() as i32, side_inflation.to_nearest_px() as i32);
+            size = inflated_size.size;
```
Comment (pcwalton) [inline components/gfx/paint_context.rs]: If you're just using the size part of the rect, why use a rect and `inflate()` at all? Isn't `size = Size2D((side_inflation.to_nearest_px() * 2) as i32, (side_inflation.to_nearest_px() * 2) as i32` enough?

**Comment by pcwalton** [comment]
Looks good to me. Squash away!

**Comment by Adenilson** [comment]
Done.

**Pull request merged**